### PR TITLE
[On Hold] Update/attachment by attachment acl configuration

### PR DESCRIPTION
**Pull request by lyx**
## Description

This PR adds the capability for users to control public file access on an attachment by attachment basis. This builds upon #1948.

Users can restrict attachment accessibility by adding the post meta key `_restrict_attachment` with a value of `true` or `1` to an attachment.

Logged in users with the `edit_posts` capability are able to access restricted attachments.

## Changelog Description

Attachments with the post meta key `_restrict_attachment` are publicly inaccessible and accessible to logged in users with the `edit_posts` capability.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [ ] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Steps to Test

To Be Added.

### CODE REVIEW

File: files/acl/restrict-unpublished-files.php [hunk 1]
```diff
@@ -38,6 +38,30 @@ function check_file_visibility( $file_visibility, $file_path ) {
 		return FILE_IS_PUBLIC;
 	}
 
+	$attachment_restrict_file_meta = get_post_meta($attachment_id, '_restrict_attachment', true);
+
+	if ( ! empty($attachment_restrict_file_meta)) {
+		
+		if (in_array( $attachment_restrict_file_meta, [true, 'true', 'yes', 1, '1'])) {
```
Comment (GaryJones) [inline files/acl/restrict-unpublished-files.php:45]: Use `true` as the third argument to `in_array()` (makes it a strict check).

**Review by GaryJones — COMMENTED** [review]
You'll want to run PHPCS over this file, as the [standards](https://github.com/Automattic/vip-go-mu-plugins/blob/71ad5e3c192e219f05c1a67b9cacf6263fe08c89/phpcs.xml.dist#L65) include WordPress-Extra, so the whitespacing would matter.

File: files/acl/restrict-unpublished-files.php (hunk 1 above)
Comment (lyx) [reply files/acl/restrict-unpublished-files.php:45]: I had this as a strict check previously - however, I experienced inconsistent conditional evaluation of the value. 🤔 

I'm going to run through tests again to see if there is some pattern of behavior and report back.

**Comment by lyx** [comment]
Unable to test on Go sandbox. Used [testing environment](https://admin.wpvip.com/#/sites/4974/cseng-wrypanda.go-vip.net).

Attempted to run curl:

`curl -IL 'http://cseng-wrypanda.go-vip.net/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void.gif'` - Received 403 Forbidden status codes for the endpoint-check-file-acl.php file. 

Seems there is something blocking curl requests to the `endpoint-check-file-acl.php`. Skipped testing on a Go Sandbox for now.

---

Will add changes suggested by @GaryJones - will retest on local dev-env.

File: files/acl/restrict-unpublished-files.php [hunk 2]
```diff
@@ -38,6 +38,30 @@ function check_file_visibility( $file_visibility, $file_path ) {
 		return FILE_IS_PUBLIC;
 	}
 
+	$attachment_restrict_file_meta = get_post_meta($attachment_id, '_restrict_attachment', true);
+
+	if ( ! empty($attachment_restrict_file_meta)) {
```
[the remark this answers is not in the document]
Comment (lyx) [reply files/acl/restrict-unpublished-files.php:43]: Makes sense...I can use more specific checks based on the data type. 

Is the main goal here readability? Or stricter typing definitions?

**Comment by WPprodigy** [comment]
> by adding the post meta key _restrict_attachment with a value of true or 1 to an attachment.

> in_array( $attachment_restrict_file_meta, [true, 'true', 'yes', 1, '1']

Since this is a new feature, can we just pick one? Shouldn't need to have to do so much type checking if we pick an official value that will be set. Other values would be invalid.

File: files/acl/restrict-unpublished-files.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (GaryJones) [reply files/acl/restrict-unpublished-files.php:43]: "Clarity of intent" would be the angle I would go for here. Appreciating exactly what `get_post_meta()` could return for different situations, and being explicit about handling those.

**Comment by lyx** [comment]
# Test Utils

## Create Unattached Media

`wp media import --title="Attachment" --porcelain "http://s0.wp.com/i/void.gif"`

Output - $UNATTACHED_IMAGE_ID

- 3

## Create Post

`wp post create --post_type=post --post_status=(draft | publish) --post_title=("Draft Post" | "Published Post") --porcelain`

Output - $(DRAFT | PUBLISH)_POST_ID

- DRAFT | 5
- PUBLISH | 7

## Attach Media to Post

`wp media import --post_id={$(DRAFT | PUBLISH)_POST_ID} --title="Attachment on (Draft | Published) Post" --porcelain "http://s0.wp.com/i/void.gif"`

Output - $(DRAFT | PUBLISH)_IMAGE_ID

- DRAFT | 6
- PUBLISH | 8

## Get Media GUID

`wp post get {$(DRAFT | PUBLISH | UNATTACHED)_IMAGE_ID} --field=guid`

Output - $(DRAFT | PUBLISH | UNATTACHED)_IMAGE_URL

- UNATTACHED | http://test-files-acl.vipdev.lndo.site/wp-content/uploads/2022/02/void.gif
- DRAFT | http://test-files-acl.vipdev.lndo.site/wp-content/uploads/2022/02/void-1.gif
- PUBLISH | http://test-files-acl.vipdev.lndo.site/wp-content/uploads/2022/02/void-2.gif

## Curl Requests

### Media Curl Request (No Auth)

#### - Unattached Media

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void.gif'`

#### - Media attached to Draft Post

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void-1.gif'`

#### - Media attached to Published Post

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void.gif-2'`

### Media Curl Request (Auth)

#### - Unattached Media

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void.gif' -H 'Cookie: wordpress_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|e4296dfb2b2920accd31364324fa1347170e834b164158d9e814025b1cc56831; wordpress_test_cookie=WP Cookie check; wordpress_logged_in_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|62803033def50067962aed2b8d7f22e8c77770a67f7c14b8c278ad7e1987dfab; wp-settings-time-1=1645194804'`

#### - Media attached to Draft Post

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void-1.gif' -H 'Cookie: wordpress_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|e4296dfb2b2920accd31364324fa1347170e834b164158d9e814025b1cc56831; wordpress_test_cookie=WP Cookie check; wordpress_logged_in_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|62803033def50067962aed2b8d7f22e8c77770a67f7c14b8c278ad7e1987dfab; wp-settings-time-1=1645194804'`

#### - Media attached to Published Post

`curl -IL 'http://test-files-acl.vipdev.lndo.site/wp-content/mu-plugins/files/acl/endpoint-check-file-acl.php' -H 'X-Original-Uri: /wp-content/uploads/2022/02/void-2.gif' -H 'Cookie: wordpress_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|e4296dfb2b2920accd31364324fa1347170e834b164158d9e814025b1cc56831; wordpress_test_cookie=WP Cookie check; wordpress_logged_in_6258e172399227b93ba1169f5ec23527=vipgo|1645367603|CNUN623qyZXoUhNAwDPlvngtnVuNzxytpHMFTsfshh9|62803033def50067962aed2b8d7f22e8c77770a67f7c14b8c278ad7e1987dfab; wp-settings-time-1=1645194804'`

## Test Cases

### 1 - No configuration

#### Parameters
- None

#### Expected Output

- Unauthenticated User
    - Response Code 202
- Authenticated User
    - Response Code 202

#### Result

Unauthenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:30:46 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:34:30 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

### 2 - Option Restrict All Enabled

#### Parameters
- Option `vip_files_acl_restrict_all_enabled` set to 1

#### Expected Output

- Unauthenticated User
    - Response Code 202
- Authenticated User
    - Response Code 202

#### Result

Unauthenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:36:11 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:36:36 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

### 3 - Option Restrict Unpublished Enabled

#### Parameters
- Option `vip_files_acl_restrict_unpublished_enabled` set to 1

#### Expected Output

- Unauthenticated User
    - Response Code 202
- Authenticated User
    - Response Code 202

#### Result

Unauthenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:39:08 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:39:08 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

### 4 - VIP_FILES_ACL_ENABLED set to true

#### Parameters
- `define('VIP_FILES_ACL_ENABLED', true)`

#### Expected Output

- Unauthenticated User
    - Response Code 202
- Authenticated User
    - Response Code 202

#### Result

Unauthenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:43:42 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:43:42 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```


### 5 - Files ACL Enabled + Restrict All Enabled

#### Parameters
- `define('VIP_FILES_ACL_ENABLED', true)`
- Option `vip_files_acl_restrict_all_enabled` set to 1

#### Expected Output

- Unauthenticated User
    - Response Code 403 
- Authenticated User
    - Response Code 202

#### Results

Unauthenticated User

```
HTTP/1.1 403 Forbidden
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:44:36 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Authenticated User

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 14:44:38 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

### 6 - Files ACL Enabled + Restrict Unpublished Enabled

#### Parameters
- `define('VIP_FILES_ACL_ENABLED', true)`
- Option `vip_files_acl_restrict_unpublished_enabled` set to 1

#### Expected Output

- Unauthenticated User (UNATTACHED POST)
    - Response Code 202
- Unauthenticated User (DRAFT POST)
    - Response Code 403 
- Unauthenticated User (PUBLISH POST)
    - Response Code 202
- Authenticated User (UNATTACHED POST)
    - Response Code 202
- Authenticated User (DRAFT POST)
    - Response Code 202
- Authenticated User (PUBLISH POST)
    - Response Code 202

#### Results

Unauthenticated User (UNATTACHED POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:02:49 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Unauthenticated User (DRAFT POST)

```
HTTP/1.1 403 Forbidden
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:02:58 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Unauthenticated User (PUBLISH POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:03:04 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User (UNATTACHED POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:03:12 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User (DRAFT POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:03:18 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Authenticated User (PUBLISH POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:03:24 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

### 7 - Files ACL Enabled + Restrict Unpublished Enabled + Attachment Meta Key Enabled

#### Parameters
- `define('VIP_FILES_ACL_ENABLED', true)`
- Option `vip_files_acl_restrict_unpublished_enabled` set to 1
- Attachment meta key `_restrict_attachment` set to 1

#### Expected Output

- Unauthenticated User (UNATTACHED POST)
    - Response Code 403
- Unauthenticated User (DRAFT POST)
    - Response Code 403 
- Unauthenticated User (PUBLISH POST)
    - Response Code 202
- Authenticated User (UNATTACHED POST)
    - Response Code 202
- Authenticated User (DRAFT POST)
    - Response Code 202
- Authenticated User (PUBLISH POST)
    - Response Code 202

#### Results

Unauthenticated User (UNATTACHED POST)

```
HTTP/1.1 403 Forbidden
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:08:57 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Unauthenticated User (DRAFT POST)

```
HTTP/1.1 403 Forbidden
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:09:05 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Unauthenticated User (PUBLISH POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:09:09 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: false
```

Authenticated User (UNATTACHED POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:09:15 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Authenticated User (DRAFT POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:09:20 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

Authenticated User (PUBLISH POST)

```
HTTP/1.1 202 Accepted
Content-Type: text/html; charset=UTF-8
Date: Fri, 18 Feb 2022 15:09:26 GMT
Server: nginx/1.21.3
X-Powered-By: PHP/7.4.27
X-Private: true
```

File: files/acl/restrict-unpublished-files.php (hunk 1 above)
Comment (lyx) [reply files/acl/restrict-unpublished-files.php]: No problems in testing. This is good to go.

**Comment by rebeccahum** [comment]
I'm seeing some ElasticPress submodule updates in this PR, can you remove that from it please?

File: files/acl/restrict-unpublished-files.php [hunk 3]
```diff
@@ -38,6 +38,30 @@ function check_file_visibility( $file_visibility, $file_path ) {
...
 	}
 
+	$attachment_restrict_file_meta = get_post_meta( $attachment_id, '_restrict_attachment', true );
+
+	// Updated conditional to use type specific empty check.
+	if ( strlen( $attachment_restrict_file_meta ) > 0 ) {
+		
+		// Configured in_array strict check.
+		// Reduced possible values to an official value to be checked.
+		if ( in_array( $attachment_restrict_file_meta, [ 1, '1' ], true ) ) {
```
Comment (jesse-greathouse) [inline files/acl/restrict-unpublished-files.php:48]: Isn't `$attachment_restrict_file_meta` truthy/falsy at this step?

Would this not be valid here?:

```php
if ( $attachment_restrict_file_meta ) {
```

**Comment by lyx** [comment]
@rebeccahum 

Alright, so what I did to address:

> I'm seeing some ElasticPress submodule updates in this PR, can you remove that from it please?

I rebased the branch locally by running: 

`git reset --hard 47819167926fe34e41d2556c6d26f1709b8fb095` - this was done to rebase the to the most recent commit before the EP submodule changes were brought in by a master merge. The original master merge was intended to bring my commit in line with the master branch of vip-go-mu-plugins.

`git push -f` - this was done to force an update to the remote branch `update/attachment-by-attachment-acl-configuration` thereby updating the PR to not include EP submodule changes.

**Comment by lyx** [comment]
@jesse-greathouse 

The following commit after this comment will utilize the falsy/truthy evaluation of the cases for the meta value.

**Comment by lyx** [comment]
Some weirdness when I tried to push after updating the conditional to use truthy/falsy evaluation. 

A `git pull` on the `update/attachment-by-attachment-acl-configuration` branch from the remote was necessary to update my local. This caused some out of date references on the remote branch to pull down more EP submodule changes.

I rebased my local again after pulling to the commit (df902b9690690549686dd88c8dc5ac767bbebca9) which represented the truthy/falsy conditional evaluation. 

After this, I once again pulled the remote branch state `update/attachment-by-attachment-acl-configuration`, which this time did not include all the extra EP submodule stuff and just registered that I had the most recent remote version of this branch.

A run of `git status` indicated that no EQ submodule code was included. Pushed the truthy/falsy evaluation successfully with issue, excluding EQ submodule fun.

**Comment by lyx** [comment]
Re-added [Status] Needs Review and leaving [Status] Has Feedback to document previous round of feedback addressed.

**Comment by lyx** [comment]
@jesse-greathouse 

Interesting, I see some of the tests are failing...I took a look at them briefly - it seems related to search...but I'm definitely nearing the end of my mental clarity for the week...so I'll take another look next week before I make a decision like that 😄 .

btw - thanks for the review! apologies for taking a while to get to it - I had thought this PR was blocked for the foreseeable  due to some P2 discussion...but maybe it can move forward? 

either way, not trying to rush pushing code - there is a workaround for WH in case this is urgent. Plus @davidsword had just mentioned to me that CDT also will need (90% change) something like this to restrict attachments.

workaround for reference - https://github.com/wpcomvip/wh46/issues/35

**Comment by jesse-greathouse** [comment]
sometimes if i get a branch tangled up with wierd meta stuff its easier just to toss the branch and start over and just copy/paste your edits into a new branch. im not sure whats going on with this one so idunno if thats the play here. no problem. always happy to have a look.

**Comment by lyx** [comment]
Removing `[Status] Needs Review` for the present time to indicate there are still a few steps I need to take.

Adding `[On Hold]` tag to PR title to notify collaborators that there are a few outstanding tasks to determine why the CI pipeline tests are failing after recent updates.

**Pull request closed without merging**